### DISCUSSION TITLE
customChange and liquibase code generation

### DISCUSSION THREAD

**Discussion by sergeymeshkov** [jOOQ/jOOQ#14797] [Q&A] [answered]
Hi, I have two 2 changelogs:
The first one creates a database table and the second one makes custom changes to this table:

```
    <changeSet id="table" author="admin">
        <createTable tableName="demo_table">
            <column name="id" type="int" autoIncrement="true">
                <constraints primaryKey="true" nullable="false"/>
            </column>
            <column name="value_one" type="text"/>
            <column name="value_two" type="int"/>
        </createTable>
    </changeSet>

```

```
 <changeSet id="custom" author="admin">
        <customChange class="org.example.CustomTask"/>
 </changeSet>
```

```
  <include file="0.1.0/master.xml" relativeToChangelogFile="true" />
  <include file="0.2.0/master.xml" relativeToChangelogFile="true" />
```
When I try to generate JOOQ code I get an error

`Caused by: liquibase.exception.ChangeLogParseException: liquibase.exception.SetupException: liquibase.parser.core.ParsedNodeException: liquibase.exception.CustomChangeException: java.lang.ClassNotFoundException: org.example.CustomTask`

I use `org.jooq:jooq-meta-extensions-liquibase` and `nu.studer.jooq` for code generation. Full sample:

```
plugins {
    id 'java'
    id 'nu.studer.jooq' version '8.1'
}

group 'org.example'
version '1.0-SNAPSHOT'

repositories {
    mavenCentral()
}

dependencies {
    compileOnly "org.jooq:jooq-meta-extensions-liquibase:3.18.0"
    jooqGenerator("org.jooq:jooq-meta-extensions-liquibase")

    implementation "org.jooq:jooq:3.18.0"
    implementation 'org.liquibase:liquibase-core:4.19.0'
    implementation 'org.postgresql:postgresql:42.4.3'

    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.8.1'
    testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.8.1'
}

jooq {
    version = '3.18.0'
    configurations {
        main {
            generateSchemaSourceOnCompilation = false

            generationTool {
                generator {
                    database {
                        name = 'org.jooq.meta.extensions.liquibase.LiquibaseDatabase'
                        properties {
                            property {
                                key = 'rootPath'
                                value = "${projectDir}/src/main/resources"
                            }
                            property {
                                key = 'scripts'
                                value = 'db/changelog/upgrade/changelog-master.xml'
                            }
                        }
                    }
                    target {
                        packageName = 'org.example'
                        directory = 'build/generated-src/jooq/main'
                    }
                }
            }
        }
    }
}

compileJava.finalizedBy generateJooq

test {
    useJUnitPlatform()
}

```

Looks like the liquibase code generator can't get access to the classpath - as a result, it can't find add necessary migrations and generate code.

**The question is:**
1. Can we somehow point the liquibase code generator to the `customChange` class? 
2. If not - can we somehow skip such changelogs? 

Thanks!

**Comment by lukaseder** [comment, marked as answer]
Well, you didn't add the project containing that class to your `dependencies`. It must be compiled and made available before you generate jOOQ code. The dependency graph is similar to that of the `JPADatabase`:
https://www.jooq.org/doc/latest/manual/code-generation/codegen-jpa/#how-to-organise-your-dependencies

```
                        +-------------------+
                        | Liquibase tasks   |
                        +-------------------+
                             ^         ^
                  depends on |         | depends on
                             |         |
          +---------------------+   +---------------------+
          | jOOQ codegen plugin |   | Your application    |
          +---------------------+   +---------------------+
                             |         |
                   generates |         | depends on
                             v         v
                     +-------------------------+
                     | jOOQ generated classes  |
                     +-------------------------+
```

I hope this helps.

**Reply by sergeymeshkov** [reply]
Hi @lukaseder my project contains the task class itself. As you can notice - the task class is compiled before JOOQ code generation:
`compileJava.finalizedBy generateJooq`
Are you saying that I somehow should add my project as a dependency itself? 
My project structure:

<img width="359" alt="Screenshot 2023-03-15 at 16 26 08" src="https://user-images.githubusercontent.com/10210266/225322343-8fb2633d-7880-4f3e-b493-eb4cb1aac372.png">

**Reply by lukaseder** [reply]
I'm sorry, I don't know gradle well enough... My second guess would have been to make it available explicitly by adding it via `jooqGenerator`, because jOOQ's code generator can't discover it automatically. Perhaps, you can ask this question on https://stackoverflow.com/questions/tagged/jooq as well? A dependency seems to be the cleanest option in my opinion, but perhaps there are more gradle idiomatic ways?

**Reply by sergeymeshkov** [reply]
Indeed jOOQ's code generator can't discover our classes automatically. As you mentioned the only way is to add it to the `jooqGenerator`. It's tricky since the task is located in the project itself. 
The simplest way to make it work is to create a new subproject with `CustomTask` and add it as a dependency.
This is an example of a custom generator, however, the same concept can be applied to the custom task - https://github.com/etiennestuder/gradle-jooq-plugin/tree/main/example/configure_custom_generator_strategy 
Thanks, @lukaseder
